harness/determined/common/api/authentication.py

```diff
@@ -62,139 +69,79 @@ def default_load_user_password(
 
-    TODO(MLG-215): delete Authentication class and write a function that returns a UsernameTokenPair
-    in its place, and let do_request() take UsernameTokenPair as input.
+    This is the login path for nearly all user-facing cases.
     """
 
-    def __init__(self, session_user: str, session_token: str) -> None:
-        self.session_user = session_user
-        self.session_token = session_token
+    master_address = master_address or util.get_default_master_address()
+    token_store = TokenStore(master_address)
 
-    def get_session_user(self) -> str:
-        return self.session_user
+    # Get session_user and password given the following priority:
+    # 1. User passed in with flag (requested_user)
+    # 2. User from environment if DET_PASS is set.
+    # 3. Active user from the token store.
+    user, password = default_load_user_password(requested_user, password, token_store)
+
+    # For login, we allow falling back to the default username.
+    if not user:
+        user = constants.DEFAULT_DETERMINED_USER
+    assert user is not None
+
+    # Check the token store if this session_user has a cached token. If so, check with the
+    # master to verify it has not expired. Otherwise, let the token be None.
+    token = token_store.get_token(user)
+    if token is not None and not _is_token_valid(master_address, token, cert):
+        token_store.drop_user(user)
+        token = None
+
+    if token is not None:
+        return UsernameTokenPair(user, token)
+
+    # Special case: use token provided from the container environment if:
+    # - No token was obtained from the token store already,
+    # - There is a token available from the container environment, and
+    # - No user was explicitly requested, or the requested user matches the token available in the
+    #   container environment.
+    if (
+        util.get_det_username_from_env() is not None
+        and util.get_det_user_token_from_env() is not None
+        and requested_user in (None, util.get_det_username_from_env())
+    ):
+        user = util.get_det_username_from_env()
+        assert user
+        token = util.get_det_user_token_from_env()
+        assert token
+        return UsernameTokenPair(user, token)
+
+    # We'll need to create a new token, so we'll need a password. If there was no requested user and
+    # we ended up falling back to the default username `determined`, then we can fall back to the
+    # default login as well. Otherwise, ask the user for their password.
+    fallback_to_default = password is None and user == constants.DEFAULT_DETERMINED_USER
+    if fallback_to_default:
+        password = constants.DEFAULT_DETERMINED_PASSWORD
+
+    if password is None:
+        password = getpass.getpass("Password for user '{}': ".format(user))
+
+    try:
+        utp = login(master_address, user, password, cert)
+        user, token = utp.username, utp.token
+    except api.errors.ForbiddenException:
+        if fallback_to_default:
```

Comment:
The differentiation of the two exception types is unclear to me. Do you think it's still helpful? Is it worth a comment?

Reply:
Thanks. I think the goal of throwing this exception here makes sense.

The logic, we probably agree, sucks. It's definitely pinging the "this is more complicated than it feels like should be" intuition bell in my head that I rely on to tell me whether I should refactor something. But it could be that nothing simpler is possible without changing some of the login fallback logic. I haven't thought about it deeply enough to say. Thinking is hard.

By the by, could you please add docstrings for these two error types in `errors.py` explaining what you found?

Reply:
Here's my opinion:

1. `errors.py` should have better docstrings per above
2. `login_with_cache` should not catch and re-throw the exception. I don't think an `UnauthorizedException` makes sense here, and the frontend should not be the decider between this class and `ForbiddenException`.

Reply:
wiki sounds great.